### DISCUSSION TITLE
Deserializers composition

### DISCUSSION THREAD

**Discussion by xamgore** [jonasbb/serde_with#664] [Q&A] [answered]
Let's say I have a struct `A`, which can be deserialized from `u8`.

```rust
#[repr(u8)]
#[derive(Deserialize_repr)]
enum A {
  Variant = 12,
}
```

And I have JSON data, where the field is presented as an integer in a string: 

```json
{
  "field": "12"
}
```

I'd like to extract the field through the sequence of deserializers: `str -> u8` & `u8 -> A`. How can I do it? (We can assume `Serialize_repr` is also implemented)

```rust
#[serde_as]
#[derive(Deserialize)]
pub struct Data {
  #[serde_as(as = "...")]
  field: A,
}
```

The best approach I could find was `JsonString<Same>`, which is a bit off.

```rust
#[serde_as(as = "JsonString<_>")]
field: A
```

**Comment by jonasbb** [comment, marked as answer]
Why are you deriving a wrong `Deserialize` implementation in the first place? If you want the enum to serialize into strings with decimal numbers, the `Deserialize` implementation should match that. The normal `serde_derive` is enough to achieve that:

```rust
#[repr(u8)]
#[derive(serde::Serialize)]
enum A {
    #[serde(rename = "12")]
    Variant = 12,
}
```

If you don't like the way `JsonString` is implemented, you could write something yourself. There is nothing in the `serde_with` crate, that will allow you to chain deserializing behavior like `str -> u8` & `u8 -> A`.

**Reply by xamgore** [reply]
Hmm, that's nice, thanks! 👍🏻 Never had an idea of renaming variants.
